zonec.c

```diff
@@ -753,6 +753,426 @@ zparser_conv_nsec(region_type *region,
+			return SVCB_KEY_ECHCONFIG;
+		break;
+
+	case sizeof("alpn")-1:
+		if (!strncmp(key, "alpn", sizeof("alpn")-1))
+			return SVCB_KEY_ALPN;
+		if (!strncmp(key, "port", sizeof("port")-1))
+			return SVCB_KEY_PORT;
+		break;
+
+	case sizeof("no-default-alpn")-1:
+		if (!strncmp( key  , "no-default-alpn"
+		            , sizeof("no-default-alpn")-1))
+			return SVCB_KEY_NO_DEFAULT_ALPN;
+		break;
+
+	case sizeof("ipv4hint")-1:
+		if (!strncmp(key, "ipv4hint", sizeof("ipv4hint")-1))
+			return SVCB_KEY_IPV4HINT;
+		if (!strncmp(key, "ipv6hint", sizeof("ipv6hint")-1))
+			return SVCB_KEY_IPV6HINT;
+		break;
+	default:
+		break;
+	}
+	if (key_len > sizeof(buf) - 1)
+		zc_error_prev_line("Unknown SvcParamKey");
+	else {
+		memcpy(buf, key, key_len);
+		buf[key_len] = 0;
+		zc_error_prev_line("Unknown SvcParamKey: %s", buf);
+	}
+	/* Although the returned value might be used by the caller,
+	 * the parser has erred, so the zone will not be loaded.
+	 */
+	return -1;
+}
+
+static uint16_t *
+zparser_conv_svcbparam_port_value(region_type *region, const char *val)
+{
+	unsigned long int port;
+	char *endptr;
+	uint16_t *r;
+
+	port = strtoul(val, &endptr, 10);
+	if (endptr > val	/* digits seen */
+	&& *endptr == 0		/* no non-digit chars after digits */
+	&&  port <= 65535) {	/* no overflow */
+
+		r = alloc_rdata(region, 3 * sizeof(uint16_t));
+		r[1] = htons(SVCB_KEY_PORT);
+		r[2] = htons(sizeof(uint16_t));
+		r[3] = htons(port);
+		return r;
+	}
+	zc_error_prev_line("Could not parse port SvcParamValue: \"%s\"", val);
+	return alloc_rdata_init(region, "", 0);
+}
+
+static uint16_t *
+zparser_conv_svcbparam_ipv4hint_value(region_type *region, const char *val)
+{
+	uint16_t *r;
+	int count;
+	char ip_str[INET_ADDRSTRLEN];
+	char *next_ip_str;
+	uint32_t *ip_wire_dst;
+	size_t i;
+
+	for (i = 0, count = 1; val[i]; i++) {
+		if (val[i] == ',')
+			count += 1;
+	}
+
+	/* count == number of comma's in val + 1 
+	 * so actually the number of IPv4 addresses in val
+	 */
+	r = alloc_rdata(region, 2 * sizeof(uint16_t) + IP4ADDRLEN * count);
+	r[1] = htons(SVCB_KEY_IPV4HINT);
+	r[2] = htons(IP4ADDRLEN * count);
+	ip_wire_dst = (void *)&r[3];
+
+	while (count) {
+		if (!(next_ip_str = strchr(val, ','))) {
+			if (inet_pton(AF_INET, val, ip_wire_dst) != 1)
+				break;
+
+			assert(count == 1);
+
+		} else if (next_ip_str - val >= (int)sizeof(ip_str))
+			break;
+
+		else {
+			memcpy(ip_str, val, next_ip_str - val);
+			ip_str[next_ip_str - val] = 0;
+			if (inet_pton(AF_INET, ip_str, ip_wire_dst) != 1) {
+				val = ip_str; /* to use in error reporting below */
+				break;
+			}
+
+			val = next_ip_str + 1;
+		}
+		ip_wire_dst++;
+		count--;
+	}
+	if (count)
+		zc_error_prev_line("Could not parse ipv4hint SvcParamValue: %s", val);
+
+	return r;
+}
+
+static uint16_t *
+zparser_conv_svcbparam_ipv6hint_value(region_type *region, const char *val)
+{
+	uint16_t *r;
+	int i, count;
+	char ip6_str[INET6_ADDRSTRLEN];
```

Comment:
Add one for a terminating byte.
```suggestion
	char ip6_str[INET6_ADDRSTRLEN+1];
```